versiondb/extsort/sort.go

```diff
@@ -0,0 +1,235 @@
+			return nil, err
+		}
+	}
+
+	s.chunkWG.Wait()
+	if len(s.failures) > 0 {
+		return nil, errors.New(strings.Join(s.failures, "\n"))
+	}
+
+	streams := make([]NextFunc, len(s.chunkFiles))
+	for i, chunkFile := range s.chunkFiles {
+		if _, err := chunkFile.Seek(0, 0); err != nil {
+			return nil, err
+		}
+		var lastItem []byte
+		reader := snappy.NewReader(chunkFile)
+		streams[i] = func() ([]byte, error) {
+			item, err := readItem(reader, lastItem)
+			if err == io.EOF {
+				return nil, nil
+			}
+			lastItem = item
+			return item, err
+		}
+	}
+
+	return NewMultiWayMerge(streams, s.lesserFunc)
+}
+
+// Close closes and remove all the temporary chunk files
+func (s *ExtSorter) Close() error {
+	var err error
+	for _, chunkFile := range s.chunkFiles {
+		err1 := chunkFile.Close()
+		if err != nil {
+			err = err1
+		}
+		err1 = os.Remove(chunkFile.Name())
+		if err != nil {
+			err = err1
+		}
+	}
+	return err
+}
+
+// sortAndSaveChunk sort the chunk in memory and save to disk in order,
+// it applies delta encoding and snappy compression to the items.
+func sortAndSaveChunk(chunk [][]byte, lesserFunc LesserFunc, output *os.File) error {
+	// sort the chunk and write to file
+	sort.Slice(chunk, func(i, j int) bool {
+		return lesserFunc(chunk[i], chunk[j])
+	})
+
+	writer := snappy.NewBufferedWriter(output)
+
+	var lastItem []byte
+	for _, item := range chunk {
+		if err := writeItem(writer, item, lastItem); err != nil {
+			return err
+		}
+		lastItem = item
+	}
+	return writer.Flush()
+}
+
+// writeItem writes the payloads with delta encoding
+func writeItem(writer io.Writer, item []byte, lastItem []byte) error {
+	// both shared and non-shared length.
+	var sizeBuf [binary.MaxVarintLen64 + binary.MaxVarintLen64]byte
+
+	shared := diffOffset(lastItem, item)
+	nonShared := len(item) - shared
+
+	n1 := binary.PutUvarint(sizeBuf[:], uint64(shared))
+	n2 := binary.PutUvarint(sizeBuf[n1:], uint64(nonShared))
+
+	if _, err := writer.Write(sizeBuf[:n1+n2]); err != nil {
+		return err
+	}
+
+	if _, err := writer.Write(item[shared:]); err != nil {
+		return err
+	}
+	return nil
+}
+
+type bufferedReader interface {
+	io.Reader
+	io.ByteReader
+}
+
+// readItem decode the payloads with delta encoding
+func readItem(
+	reader bufferedReader,
+	lastItem []byte,
+) ([]byte, error) {
+	shared, err := binary.ReadUvarint(reader)
+	if err != nil {
+		return nil, err
+	}
+	nonShared, err := binary.ReadUvarint(reader)
+	if err != nil {
+		if err == io.EOF && nonShared > 0 {
+			return nil, io.ErrUnexpectedEOF
+		}
+		return nil, err
+	}
+
+	item := make([]byte, shared+nonShared)
+	copy(item[:shared], lastItem[:shared])
+	if _, err := io.ReadFull(reader, item[shared:]); err != nil {
+		return nil, err
+	}
+	return item, nil
+}
+
+// diffOffset returns the index of first byte that's different in two bytes slice,
+// the index is encoded as uint8, so the max supported length is `math.MaxUint8`,
+// that should be more than enough in our use case.
```

Comment:
if we assume max supported length is `math.MaxUint8`. maybe do panic if we see the length exceeds the max?

Reply:
out of date comment, it use varint encoding, so support int64 technically.